src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/BatchableWorkItem.cs

```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Razor.Workspaces
+{
+    /// <summary>
+    /// A work item that represents a unit of work. This is intended to be overridden so consumers can represent any
+    /// unit of work that fits their need.
+    /// </summary>
+    internal abstract class BatchableWorkItem
```

Comment:
Would this perhaps make more sense as an interface? Usually when I think of abstract classes I assume there's some sort of implementation inside the class

Reply:
Good question! Funny enough I'm pretty against interfaces all together 😄. A few of my personal reasons:

1. They encourage violation of the single responsibility principal and can easily result in god-classes/poorly factored code
2. Older versions of .NET (default interface impls fixed this in newer versions) they're less flexible than abstract classes which results in unexpected breaks / creation of multiple interfaces. Think VS where there's `ITextBuffer`, `ITextBuffer2` etc., these are because things break external consumers.
3. Similar to bullet point 1 They invoke an "I can" philosophy versus an "I am" philosophy. Aka:
```C#
interface Drink { get; set; }
interface Food { get; set; }
// This is possible with interfaces, if they were abstract classes it'd force you to pick one (a good thing)
class Steak : Food, Drink {}
```
4. Abstract classes when written as contracts (no shared methods in them) prevent inheritance. I view this as a good thing. Very rarely have I found value in having base class functionality on things outside of tests. In product code I'd rather build services / helpers that embody a single responsibility then build out functional base classes. I'm a bit of a hypocrit saying this one because we violate this in [a few](https://github.com/dotnet/aspnetcore-tooling/blob/8e66e506548205fce25814311c677fd9a9e92f8e/src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/VirtualDocumentFactoryBase.cs#L60-L101) places in Razor. One day i'll go back and fix that 😄.
  1. There's a whole bunch of unsaid sub-bullets to avoiding inheritance IMO. Gist being I've personally found inheritance makes code unreadable, harder to navigate and harder to reason about.

Now all that being said, there are loads of people who feel very different from I on interfaces but this is the philosophy I embrace. Not saying I'm right, just sharing my reasoning 😄

Reply:
Could avoid this whole conversation and just have this be a delegate 😛

Reply:
> Could avoid this whole conversation and just have this be a delegate 😛

I had plans to expand the work item to have more methods/capabilities 😄, you right though, in its current form a delegate is sufficient

Reply:
Thanks for the detailed write up! Makes sense, you make great points and have me almost convinced against using interfaces too 😄